From: Nathaniel Abalos <ncabalos@truefitness.com>
Date: Tue, 26 Aug 2014 13:09:20 -0500
Subject: added separate list for browser handlers

---
 audio/avctp.c |    8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/audio/avctp.c b/audio/avctp.c
index 9ba1cfc..b417fee 100644
--- a/audio/avctp.c
+++ b/audio/avctp.c
@@ -155,6 +155,7 @@ struct avctp {
 
 	uint8_t key_quirks[256];
 	GSList *handlers;
+	GSList *browsing_handlers;
 	
 	bool initiator;
 };
@@ -385,6 +386,7 @@ static void avctp_disconnected(struct avctp *session)
 	server = session->server;
 	server->sessions = g_slist_remove(server->sessions, session);
 	g_slist_free_full(session->handlers, g_free);
+	g_slist_free_full(session->browsing_handlers, g_free);
 	g_free(session);
 }
 
@@ -577,7 +579,7 @@ static void browsing_response(struct avctp *session, struct avctp_header *avctp,
 {
 	GSList *l;
 	
-	for (l = session->handlers; l; l = l->next) {
+	for (l = session->browsing_handlers; l; l = l->next) {
 		struct avctp_browsing_rsp_handler *handler = l->data;
 		
 		if (handler->id != avctp->transaction)
@@ -588,7 +590,7 @@ static void browsing_response(struct avctp *session, struct avctp_header *avctp,
 					handler->user_data))
 				return;
 				
-		session->handlers = g_slist_remove(session->handlers, handler);
+		session->browsing_handlers = g_slist_remove(session->browsing_handlers, handler);
 		g_free(handler);
 			
 		return;			
@@ -1255,7 +1257,7 @@ int avctp_send_browsing_req(struct avctp *session,
 	handler->func = func;
 	handler->user_data = user_data;
 
-	session->handlers = g_slist_prepend(session->handlers, handler);
+	session->browsing_handlers = g_slist_prepend(session->browsing_handlers, handler);
 	
 	id++;
 	id %= 16; //transaction id is only 4 bits
